### ISSUE TITLE
[Code Question] what are those underscore mean in c++ 11

### ISSUE THREAD

**Issue by baixianger** [sail-sg/envpool#268]
I was reading the code from dummy env and am not a good C++er. So I have many questions when reading your code.

1. What are those underscores when indexing a map or dict?
```
      state["info:players.id"_][i] = i;
      state["info:players.done"_][i] = IsDone();
      state["obs:raw"_](i, 0) = state_;
      state["obs:raw"_](i, 1) = 0;
      state["reward"_][i] = -i;
```

2. What do those underscores mean>, BTW, where do those `.Bind()` functions come from?
```
  template <typename Config>
  static decltype(auto) ActionSpec(const Config& conf) {
    return MakeDict("list_action"_.Bind(Spec<double>({6})),
                    "players.action"_.Bind(Spec<int>({-1})),
                    "players.id"_.Bind(Spec<int>({-1})));
  }
```

**Comment by Trinkle23897** [comment]
underscore: https://github.com/sail-sg/envpool/blob/aacf06f694ead2eb75331f085f00dad71eec1a08/envpool/core/dict.h#L58-L60

tldr: in compilation time, it will convert the string to an ID to check the key-value pair in this class

Bind: https://github.com/sail-sg/envpool/blob/aacf06f694ead2eb75331f085f00dad71eec1a08/envpool/core/dict.h#L51-L53

See examples here: https://github.com/sail-sg/envpool/blob/main/envpool/core/dict_test.cc

**Comment by baixianger** [comment]
> underscore:
> 
> https://github.com/sail-sg/envpool/blob/aacf06f694ead2eb75331f085f00dad71eec1a08/envpool/core/dict.h#L58-L60
> 
> tldr: in compilation time, it will convert the string to an ID to check the key-value pair in this class
> 
> Bind:
> 
> https://github.com/sail-sg/envpool/blob/aacf06f694ead2eb75331f085f00dad71eec1a08/envpool/core/dict.h#L51-L53
> 
> See examples here: https://github.com/sail-sg/envpool/blob/main/envpool/core/dict_test.cc

Thank you for being so helpful.

**Comment by baixianger** [comment]
> underscore:
> 
> https://github.com/sail-sg/envpool/blob/aacf06f694ead2eb75331f085f00dad71eec1a08/envpool/core/dict.h#L58-L60
> 
> tldr: in compilation time, it will convert the string to an ID to check the key-value pair in this class
> 
> Bind:
> 
> https://github.com/sail-sg/envpool/blob/aacf06f694ead2eb75331f085f00dad71eec1a08/envpool/core/dict.h#L51-L53
> 
> See examples here: https://github.com/sail-sg/envpool/blob/main/envpool/core/dict_test.cc

作者您好，今天是读您源代码学习的第二天。 看到您这里自己实现了一个dict的功能，然后就想咨询一下，您的这个dict和stl里面实现的map数据类型有啥区别和优势？ 非常谢谢。
